[deckhouse-controller] deckhouse throws panics on restart if there any pending module releases in the cluster, waiting for being deployed

### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Version

v1.56

### Expected Behavior

Pending module releases should be scheduled to be deployed only after deckhouse-config has been initialized.

### Actual Behavior

If there are modules releases in pending state in the cluster, on deckhouse restart, deckhouse-controller tries to deploy these releases faster than the deckhouse-config structure gets initialized, causing a panic.

### Steps To Reproduce

1. create module release with such an update policy that would try to deploy release in 5 minutes
2. stop deckhouse
3. wait for the module release update window
4. start deckhouse

### Additional Information

_No response_

### Logs

```shell
{"level":"error","msg":"E0115 11:27:40.149409      14 runtime.go:77] Observed a panic: deckhouse-config Service is not initialized\ngoroutine 178 [running]:\nk8s.io/apimachinery/pkg/util/runtime.logPanic({0x3bff280?, 0x486ef20})\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/runtime/runtime.go:75 +0x85\nk8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000e3c790?})\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/runtime/runtime.go:49 +0x6b\npanic({0x3bff280?, 0x486ef20?})\n\t/usr/local/go/src/runtime/panic.go:914 +0x21f\ngithub.com/deckhouse/deckhouse/go_lib/deckhouse-config.Service(...)\n\t/deckhouse/go_lib/deckhouse-config/service.go:54\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).reconcilePendingRelease(0xc000cfe400, {0x48a8a78, 0x64e61a0}, 0xc001debe00)\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:438 +0x22d4\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).createOrUpdateReconcile(0xc000cfe400, {0x48a8a78, 0x64e61a0}, 0xc000d72a80?)\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:400 +0x6d7\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).Reconcile(0xc000cfe400, {0x48a8a78, 0x64e61a0}, {0xc000cf37c0?, 0x0?})\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:624 +0xc8\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).processNextWorkItem.func1(0xc000cfe400, {0x48a8a78?, 0x64e61a0?}, {0x3bff280?, 0xc000e3c790?})\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:278 +0xfd\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).processNextWorkItem(0xc000cfe400, {0x48a8a78, 0x64e61a0})\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:292 +0x65\ngithub.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).runWorker(...)\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:242\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:259 +0x22\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:226 +0x33\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x487bd80, 0xc001e84240}, 0x1, 0x0)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:227 +0xaf\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:204 +0x7f\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x48a8a78, 0x64e61a0}, 0xc0012c8fb0, 0x0?, 0x0?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:259 +0x93\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x48a8a78?, 0x64e61a0?}, 0x0?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:170 +0x25\ncreated by github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).Run in goroutine 120\n\t/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:234 +0x37a\n","source":"klog","time":"2024-01-15T11:27:40Z"}
panic: deckhouse-config Service is not initialized [recovered]
	panic: deckhouse-config Service is not initialized

goroutine 178 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000e3c790?})
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/runtime/runtime.go:56 +0xcd
panic({0x3bff280?, 0x486ef20?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/deckhouse/deckhouse/go_lib/deckhouse-config.Service(...)
	/deckhouse/go_lib/deckhouse-config/service.go:54
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).reconcilePendingRelease(0xc000cfe400, {0x48a8a78, 0x64e61a0}, 0xc001debe00)
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:438 +0x22d4
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).createOrUpdateReconcile(0xc000cfe400, {0x48a8a78, 0x64e61a0}, 0xc000d72a80?)
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:400 +0x6d7
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).Reconcile(0xc000cfe400, {0x48a8a78, 0x64e61a0}, {0xc000cf37c0?, 0x0?})
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:624 +0xc8
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).processNextWorkItem.func1(0xc000cfe400, {0x48a8a78?, 0x64e61a0?}, {0x3bff280?, 0xc000e3c790?})
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:278 +0xfd
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).processNextWorkItem(0xc000cfe400, {0x48a8a78, 0x64e61a0})
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:292 +0x65
github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).runWorker(...)
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:242
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:259 +0x22
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x487bd80, 0xc001e84240}, 0x1, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x48a8a78, 0x64e61a0}, 0xc0012c8fb0, 0x0?, 0x0?, 0x0?)
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:259 +0x93
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x48a8a78?, 0x64e61a0?}, 0x0?, 0x0?)
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/wait/backoff.go:170 +0x25
created by github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release.(*Controller).Run in goroutine 120
	/deckhouse/deckhouse-controller/pkg/controller/module-controllers/release/controller.go:234 +0x37a
```
```

--- a/deckhouse-controller/pkg/controller/module-controllers/release/controller.go
+++ b/deckhouse-controller/pkg/controller/module-controllers/release/controller.go
@@ -217,6 +217,16 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 	defer utilruntime.HandleCrash()
 	defer c.workqueue.ShutDown()
 
+	// Check if controller's dependencies have been initialized
+	_ = wait.PollUntilContextCancel(ctx, utils.SyncedPollPeriod, false,
+		func(context.Context) (bool, error) {
+			// TODO: add modulemanager initialization check c.modulesValidator.AreModulesInited() (required for reloading modules without restarting deckhouse)
+			if deckhouseconfig.IsServiceInited() {
+				return true, nil
+			}
+			return false, nil
+		})
+
 	// Start the informer factories to begin populating the informer caches
 	c.logger.Info("Starting ModuleRelease controller")
 

--- a/deckhouse-controller/pkg/controller/module-controllers/release/controller.go
+++ b/deckhouse-controller/pkg/controller/module-controllers/release/controller.go
@@ -221,10 +221,7 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 	_ = wait.PollUntilContextCancel(ctx, utils.SyncedPollPeriod, false,
 		func(context.Context) (bool, error) {
 			// TODO: add modulemanager initialization check c.modulesValidator.AreModulesInited() (required for reloading modules without restarting deckhouse)
-			if deckhouseconfig.IsServiceInited() {
-				return true, nil
-			}
-			return false, nil
+			return deckhouseconfig.IsServiceInited(), nil
 		})
 
 	// Start the informer factories to begin populating the informer caches

--- a/deckhouse-controller/pkg/controller/module-controllers/release/override-controller.go
+++ b/deckhouse-controller/pkg/controller/module-controllers/release/override-controller.go
@@ -41,6 +41,7 @@ import (
 	d8listers "github.com/deckhouse/deckhouse/deckhouse-controller/pkg/client/listers/deckhouse.io/v1alpha1"
 	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/downloader"
 	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/controller/module-controllers/utils"
+	deckhouseconfig "github.com/deckhouse/deckhouse/go_lib/deckhouse-config"
 )
 
 // ModulePullOverrideController is the controller implementation for ModulePullOverride resources
@@ -128,6 +129,16 @@ func (c *ModulePullOverrideController) Run(ctx context.Context, workers int) {
 
 	defer utilruntime.HandleCrash()
 	defer c.workqueue.ShutDown()
+
+	// Check if controller's dependencies have been initialized
+	_ = wait.PollUntilContextCancel(ctx, utils.SyncedPollPeriod, false,
+		func(context.Context) (bool, error) {
+			// TODO: add modulemanager initialization check c.modulesValidator.AreModulesInited() (required for reloading modules without restarting deckhouse)
+			if deckhouseconfig.IsServiceInited() {
+				return true, nil
+			}
+			return false, nil
+		})
 
 	// Start the informer factories to begin populating the informer caches
 	c.logger.Info("Starting controller")

--- a/deckhouse-controller/pkg/controller/module-controllers/release/override-controller.go
+++ b/deckhouse-controller/pkg/controller/module-controllers/release/override-controller.go
@@ -134,10 +134,7 @@ func (c *ModulePullOverrideController) Run(ctx context.Context, workers int) {
 	_ = wait.PollUntilContextCancel(ctx, utils.SyncedPollPeriod, false,
 		func(context.Context) (bool, error) {
 			// TODO: add modulemanager initialization check c.modulesValidator.AreModulesInited() (required for reloading modules without restarting deckhouse)
-			if deckhouseconfig.IsServiceInited() {
-				return true, nil
-			}
-			return false, nil
+			return deckhouseconfig.IsServiceInited(), nil
 		})
 
 	// Start the informer factories to begin populating the informer caches

--- a/deckhouse-controller/pkg/controller/module-controllers/utils/utils.go
+++ b/deckhouse-controller/pkg/controller/module-controllers/utils/utils.go
@@ -15,8 +15,14 @@
 package utils
 
 import (
+	"time"
+
 	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/apis/deckhouse.io/v1alpha1"
 	"github.com/deckhouse/deckhouse/go_lib/dependency/cr"
+)
+
+const (
+	SyncedPollPeriod = 100 * time.Millisecond
 )
 
 // GenerateRegistryOptions feteches settings from ModuleSource and generate registry options from them

--- a/go_lib/deckhouse-config/service.go
+++ b/go_lib/deckhouse-config/service.go
@@ -30,7 +30,7 @@ import (
 
 var (
 	serviceInstance     *ConfigService
-	serviceInstanceLock sync.RWMutex
+	serviceInstanceLock sync.Mutex
 )
 
 func InitService(mm ModuleManager) {
@@ -50,9 +50,7 @@ func InitService(mm ModuleManager) {
 }
 
 func IsServiceInited() bool {
-	serviceInstanceLock.RLock()
-	defer serviceInstanceLock.RUnlock()
-	return serviceInstance.possibleNames != nil && serviceInstance.moduleNamesToSources != nil
+	return serviceInstance != nil
 }
 
 func Service() *ConfigService {

--- a/go_lib/deckhouse-config/service.go
+++ b/go_lib/deckhouse-config/service.go
@@ -30,7 +30,7 @@ import (
 
 var (
 	serviceInstance     *ConfigService
-	serviceInstanceLock sync.Mutex
+	serviceInstanceLock sync.RWMutex
 )
 
 func InitService(mm ModuleManager) {
@@ -47,6 +47,12 @@ func InitService(mm ModuleManager) {
 		statusReporter:       NewModuleInfo(mm, possibleNames),
 		moduleNamesToSources: make(map[string]string),
 	}
+}
+
+func IsServiceInited() bool {
+	serviceInstanceLock.RLock()
+	defer serviceInstanceLock.RUnlock()
+	return serviceInstance.possibleNames != nil && serviceInstance.moduleNamesToSources != nil
 }
 
 func Service() *ConfigService {

